lib/petree_api_web/controllers/tree_controller.ex

```diff
@@ -6,35 +6,62 @@ defmodule PetreeApiWeb.TreeController do
 
   action_fallback PetreeApiWeb.FallbackController
 
+  def index(conn, %{"user_id" => user_id}) do
+    trees = Trees.list_trees(user_id)
+    render(conn, "index.json", trees: trees)
+  end
+
   def index(conn, _params) do
     trees = Trees.list_trees()
     render(conn, "index.json", trees: trees)
   end
 
-  def create(conn, %{"tree" => tree_params}) do
-    with {:ok, %Tree{} = tree} <- Trees.create_tree(tree_params) do
+  def create(conn, params) do
+    params = extract_location(params)
+
+    with {:ok, %Tree{}} <- Trees.create_tree(params) do
       conn
-      |> put_status(:created)
-      |> put_resp_header("location", Routes.tree_path(conn, :show, tree))
-      |> render("show.json", tree: tree)
+      |> send_resp(201, "")
     end
   end
 
+  def show(conn, %{"user_id" => user_id, "tree_id" => tree_id}) do
+    tree = Trees.get_tree!(tree_id, user_id)
+    render(conn, "show.json", tree: tree)
+  end
+
   def show(conn, %{"id" => id}) do
     tree = Trees.get_tree!(id)
     render(conn, "show.json", tree: tree)
   end
 
-  def update(conn, %{"id" => id, "tree" => tree_params}) do
+  def update(conn, %{"id" => id} = params) do
+    params = extract_location(params)
+
     tree = Trees.get_tree!(id)
 
-    with {:ok, %Tree{} = tree} <- Trees.update_tree(tree, tree_params) do
+    with {:ok, %Tree{} = tree} <- Trees.update_tree(tree, params) do
       render(conn, "show.json", tree: tree)
     end
   end
 
-  def delete(conn, %{"id" => id}) do
-    tree = Trees.get_tree!(id)
+  defp extract_location(params) do
+    if Map.has_key?(params, "lat") or Map.has_key?(params, "lng") do
+      if Map.has_key?(params, "lat") and Map.has_key?(params, "lng") do
```

Comment:
I didn't get it, why you don't verify only with the second `if`?

Reply:
the extract_location function is used when a new tree is inserted or a tree is updated, using just the second "if" would work when a tree is inserted, but during an update there is no mandatory fields. that means that you can update any information that you want, using just the second "if" you are forcing the client to always send both lat and lng informations in each update. summarizing, the first "if" verify if one of field (lat or lng) is inside the request body, if one of them is inside, then the other one MUST be there too, other wise it will rise an error.